pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -0,0 +1,546 @@
+				getCacheKeyWithPrefix(ciImport.Spec.ClusterID),
+				peerConfigs,
+				tunnelPeerIP,
+				localGatewayIP); err != nil {
+				return fmt.Errorf("failed to install flows to remote Gateway in ClusterInfoImport %s: %v", ciImport.Name, err)
+			}
+		} else {
+			tunnelPeerIP := net.ParseIP(gw.InternalIP)
+			if err := c.ofClient.InstallMulticlusterNodeFlows(
+				getCacheKeyWithPrefix(gw.Name+ciImport.Spec.ClusterID),
+				peerConfigs,
+				tunnelPeerIP); err != nil {
+				return fmt.Errorf("failed to install flows to Gateway %s: %v", gw.Name, err)
+			}
+		}
+
+		c.installedCIImport.Add(&ciImportInfo{
+			localGWName: gw.Name,
+			ciImport:    *ciImport,
+		})
+	}
+	c.installedGWName = gw.Name
+	return nil
+}
+
+func (c *MCRouteController) deleteMCNodeFlowsForCIImportEvents(ciImp *mcv1alpha1.ClusterInfoImport, key string) error {
+	if err := c.ofClient.UninstallMulticlusterFlows(key); err != nil {
+		return fmt.Errorf("failed to uninstall multi-cluster flows to remote Gateway Node %s: %v", ciImp.Name, err)
+	}
+	c.installedCIImport.Delete(&ciImportInfo{ciImport: *ciImp})
+	return nil
+}
+
+func (c *MCRouteController) deleteMCNodeFlowsForGatewayEvents(gwName string) error {
+	cachedciImps := c.installedCIImport.List()
+	for _, ciImp := range cachedciImps {
+		cached := ciImp.(*ciImportInfo)
+		key := getCacheKeyWithPrefix(gwName + cached.ciImport.Spec.ClusterID)
+		if err := c.ofClient.UninstallMulticlusterFlows(key); err != nil {
+			return fmt.Errorf("failed to uninstall multi-cluster flows to remote Gateway in ClusterInfoImport %s: %v",
+				cached.ciImport.Name, err)
+		}
+		c.installedCIImport.Delete(ciImp)
+	}
+	return nil
+}
+
+func (c *MCRouteController) getAllCIImports(ciImp *mcv1alpha1.ClusterInfoImport) ([]*mcv1alpha1.ClusterInfoImport, error) {
+	var allciImports []*mcv1alpha1.ClusterInfoImport
+	var err error
+	if ciImp == nil {
+		allciImports, err = c.ciImportLister.ClusterInfoImports(c.namespace).List(labels.Everything())
+		if err != nil {
+			return nil, fmt.Errorf("failed to get ClusterInfoImport list, err: %v", err)
+		}
+		if len(allciImports) == 0 {
+			klog.InfoS("No valid remote ClusterInfo imported, do nothing")
+			return nil, nil
+		}
+	} else {
+		allciImports = append(allciImports, ciImp)
+	}
+	return allciImports, nil
+}
+
+// checkActiveGateway will compare Gateway's CreationTimestamp to get active Gateway,
+// and also check if given name is a Gateway. The last created Gateway will be the active Gateway.
+func (c *MCRouteController) checkActiveGateway() (*mcv1alpha1.Gateway, bool, error) {
+	gwlist, err := c.gwLister.Gateways(c.namespace).List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "Failed to get Gateway list", "namespace", c.namespace)
+		return nil, false, err
+	}
+	if len(gwlist) == 0 {
+		return nil, false, fmt.Errorf("no Gateway found in Namespace %s", c.namespace)
+	}
+	var iAmGW bool
+	for _, gw := range gwlist {
+		if gw.Name == c.nodeConfig.Name {
+			iAmGW = true
```

Comment:
This is strange. Do we care this Node matches any GW or not, or we just care it is/was the active Gateway (if latter, I feel this func needs not to return a boolean at all and maybe should be renamed to getActiveGateway, as it is quite easy to check Node name with the active GW name in the caller)?

Reply:
We need to check the Gateway list to understand if the current Node is a Gateway or not because it may not be the active Gateway, I have spitted this into multiple methods to make it clearer.